src/storage/mvcc/reader.rs

```diff
@@ -417,4 +418,236 @@ impl<'a> MvccReader<'a> {
             keys.push(key);
         }
     }
+
+    // Returns true if it needs gc.
+    // This is for optimization purpose, does not mean to be accurate.
+    pub fn need_gc(&self, safe_point: u64) -> bool {
+        let mut opts = GetPropertiesOptions::default();
+        opts.max_ts = Some(safe_point);
+        let props = match self.snapshot.get_properties_cf(CF_WRITE, &opts) {
+            Ok(v) => v,
+            Err(_) => return true,
+        };
+        // We don't need gc if:
+        // 1. min_ts > safe_point:
+        //    No data older than safe_point to gc.
+        if props.min_ts > safe_point {
+            return false;
+        }
+        // 2. num_keys == num_versions && num_puts == num_versions:
+        //    No keys have more than one versions, and the version is effective.
+        //    Notice: Since the properties are file-based, it can be false positive.
+        //    For example, if multiple files have a different version of the same key.
+        !(props.num_keys == props.num_versions && props.num_puts == props.num_versions)
```

Comment:
Maybe I should add a TiDB variable like `tikv_gc_min_ratio` and default to 0.1?
How do you think @disksing @AndreMouche ?